Unify linear solvers

atgeirr opened the pull request:
This is the last (I hope) take on this, replacing #2706. There are still two requests @blattms made in that PR that need to be addressed here:

> If we skip --use-cpr like in this PR then to the very least I would suggest having "cpr" here.
But "--linear-solver-configuration=cpr" is stil a lot more to type than "--use-cpr"

Part of my goal was to get rid of parameters that were redundant or came into conflict with each other. Adding "cpr" to the set of strings accepted by "--linear-solver-configuration" is no problem, just not sure if we should choose the trueimpes or quasiimpes as default (probably trueimpes, as that is what you get with --use-cpr right now). However keeping the "--use-cpr" will then conflict with the other one potentially. Would it address your concern if we shortened "--linear-solver-configuration" to "--linear-solver" or even "--linsolver"?

> How do I select AMG as preconditioner (formerly --use-amg), I think that proved good for some 2-phase problems of @totto82?

I think adding the option string "amg" to the above-discussed parameter (and an associated setup like for "ilu0" and the "cpr_*" options) should handle this. Could you advise about the actual case that benefited from it @totto82 or @blattms? Would be good to verify that the new is no worse.

atgeirr commented:
jenkins build this serial please

atgeirr commented:
benchmark please

ytelses commented:
Benchmark result overview: 




| Test    | Configuration                              | Relative |
| ------- | ------------------------------------------ | -------- |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 1 | 1.006    |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 8 | 1.005    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 1 | 0.997    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 8 | 0.992    |
* Speed-up = Total time master / Total time pull request. Above 1.0 is an improvement. *



View result details @ https://www.ytelses.com/opm/?page=result&id=922

alfbr commented:
Please note that the variance in results have been significantly reduced, performance differences at the 1% level should be detectable now. In this case, it is clear that there are no significant performance impact of this PR,

bska commented on opm/simulators/linalg/FlowLinearSolverParameters.hpp:
```diff
@@ -277,48 +237,48 @@ namespace Opm
-            cpr_reuse_setup_          = 0;
-        }
-    };
+    // struct CPRParameter
```
Do you expect to actually remove this structure,and its associate support code at some point or is this just a temporary workaround?

atgeirr replied:
There are some commented-out parts that were left in, initially to see if they could be safely removed. I'll make some cleanup.

bska reviewed (approved):
I think I'd personally write the `shouldCreateSolver()` function using early returns like below, but it's a style issue so I'm not going to insist.  Otherwise this looks good to me.

---

```C++
bool shouldCreateSolver() const
{
    // Decide if we should recreate the solver or just do
    // a minimal preconditioner update.
    if (!flexibleSolver_) {
         return true;
    }

    if (this->parameters_.cpr_reuse_setup_ == 0) {
        return true;
    }

    if (this->parameters_.cpr_reuse_setup_ == 1) {
        // Recreate solver on the first iteration of every timestep.
        return this->simulator_.model().newtonMethod().numIterations() == 0;
    }

    if (this->parameters_.cpr_reuse_setup_ == 2) {
        // Recreate solver if the last solve used more than 10 iterations.
        return this->iterations() > 10;
    }

    // Never recreate solver.
    assert(this->parameters_.cpr_reuse_setup_ == 3);

    return false;
}
```

atgeirr commented:
> write the `shouldCreateSolver()` function using early returns

I'll adopt your suggestion.

atgeirr commented:
The following changes have been made:
 - Make the changes suggested by @bska.
 - Refactor setupPropertyTree() and add new option "amg" to `--linear-solver-configuration`, as requested by @blattms.

I have tested the amg option on Norne, and it seems to work, although slower than either "ilu0" (default) or "cpr_trueimpes" for that case. It could probably benefit from some tuning to be as good as possible for the two-phase cases mentioned where it was good before, such testing should be very easy using the JSON file approach (start testing from the tree printed in the DBG file) with no recompilation. When the best defaults are found, they can be changed in setupPropertyTree.cpp.

atgeirr commented:
jenkins build this serial please

bska commented on opm/simulators/linalg/ISTLSolverEbos.hpp:
```diff
@@ -785,103 +375,74 @@ DenseMatrix transposeDenseMatrix(const DenseMatrix& M)
+            }
+            if (this->parameters_.cpr_reuse_setup_ == 1) {
+                // Recreate solver on the first iteration of every timestep.
+                const int newton_iteration = this->simulator_.model().newtonMethod().numIterations();
+                if (newton_iteration == 0) {
+                    return true;
                 }
```
Writing it this way will trigger the [`assert`](https://github.com/OPM/opm-simulators/blob/30a4fe55b060dbf940565dc4fed22a15d318cdad/opm/simulators/linalg/ISTLSolverEbos.hpp#L407) at the end of the function if `newton_iteration > 0`.  You really need an unconditional `return` here&mdash;or remove the `assert`&mdash;if you don't want that to happen.

bska commented on opm/simulators/linalg/ISTLSolverEbos.hpp:
```diff
@@ -785,103 +375,74 @@ DenseMatrix transposeDenseMatrix(const DenseMatrix& M)
-                {
-                    noGhostMat_->addindex(dofId, *nabIdx);
+            if (this->parameters_.cpr_reuse_setup_ == 2) {
+                // Recreate solver if the last solve used more than 10 iterations.
+                if (this->iterations() > 10) {
+                    return true;
                 }
```
Writing it this way will trigger the `assert` at the end of the function if `this->iteration() <= 10`. You really need an unconditional return here&mdash;or remove the `assert`&mdash;if you don't want that to happen.

atgeirr replied:
Simplified and fixed (I think).

bska replied:
> Simplified and fixed (I think).

Yep, looks good now.  I vote to merge this once the build check is green but will let others speak to this too.

bska reviewed (commented):
Thanks for the updates to `shouldCreateSolver`.  I just noticed a couple of problems with the new logic, but otherwise it looks good to me.

atgeirr commented:
jenkins build this serial please

blattms commented on tests/test_blackoil_amg.cpp:
```diff
@@ -289,30 +293,18 @@ void runBlackoilAmgLaplace()
+
+    boost::property_tree::ptree prm;
+    prm.put("type", "amg");
+    std::function<Vector()> weights = [&mat]() {
+        return Opm::Amg::getQuasiImpesWeights<BCRSMat, Vector>(mat, 0, false);
+    };
+    auto amg = Opm::PreconditionerFactory<Operator, Communication>::create(fop, prm, weights, comm);
```
This looks a bit strange and is hard to understand. At first I thought we would actually set up the weights, but the function is of course never called. Why not make that explicit?
```
auto dummy_weights = [](){return Vector();}
auto amg = Opm::PreconditionerFactory<Operator, Communication>::create(fop, prm, dummy_weights, comm);
```

blattms replied:
Don't bother. I did not see that this is just in a test.

blattms reviewed (changes requested):
Please note, that with the deletion of some of the files we might also be loosing some implemented functionality. Not sure whether this is intended (because they proofed to be bad/not working) or not, but it might be hard to recover it:
. BlackoilAMG had the option to start the second step on the first coarse level pressure sytem (now it will be only the finest one).
- I think some different scaling techniques to recover the pressure system might be lost (or were already lost previously).

There has been a vast amount of refactoring. Please make sure that proper credit is given and that copyright for copied/refactored stuff is correct.

blattms commented:
> jenkins build this serial please

Not sure: Does this build parallel, too?

akva2 commented:
yes.

blattms commented:
I have another suggestion. Why not skip the --linear-solver-configuration-file and just intrepret --linear-solver-configuration as a  file if the name matches ".*.json"?

atgeirr commented:
> I have another suggestion. Why not skip the --linear-solver-configuration-file option and just intrepret --linear-solver-configuration as a file if the name matches ".*.json"?

I like that idea. Should we also shorten it to `--linsolver`?

blattms commented:
The shorter the better, if you ask me.

atgeirr commented:
> The shorter the better, if you ask me.

I have implemented that (--linsolver). I have also removed the separate file parameter, replacing it with passing "--linsolver=options.json" as suggested. The doc string has been updated.

> There has been a vast amount of refactoring. Please make sure that proper credit is given and that copyright for copied/refactored stuff is correct.

I have added 2020 SINTEF and Equinor copyright to the ISTLSolver.hpp file. However I have not done any other additions to copyrights in the PR. Looking at the files changed by this PR, I see no other significant enough changes to warrant adding a SINTEF copyright, the rest is just removing stuff. Did you have any particular concern?

With that, I hope this can be merged when green again.

atgeirr commented:
jenkins build this serial please

atgeirr commented:
I must have made a serious mistake in one of the last commits, and it is a very embarrassing one. I'll amend ASAP...

atgeirr commented:
jenkins build this serial please

bska commented on opm/simulators/linalg/setupPropertyTree_impl.hpp:
```diff
@@ -39,7 +39,7 @@ setupPropertyTree(FlowLinearSolverParameters p) // Note: copying the parameters
 
     // Get configuration from file.
-    if (conf.substr(conf.size() - 5, 5) == ".json") {
+    if (conf.size() > 5 && conf.substr(conf.size() - 5, 5) == ".json") { // the ends_with() method is not available until C++20
```
Just out of curiosity, did you consider using `string::rfind()`?  E.g., something like
```C++
if (conf.rfind(".json") != std::string::npos) {
    // 'conf' is *.json
}
```

blattms replied:
Putting my smart glasses on and playing smart-ass:

Doesn't really cut it as it would return true for file finame.json.txt, too.

atgeirr replied:
> Just out of curiosity, did you consider using `string::rfind()`? E.g., something like

I did not. I see that would be a little easier, but it does not do exactly the same: it would find the string anywhere, not just at the end. I guess that is mostly academic though.

bska commented on opm/simulators/linalg/ISTLSolverEbos.hpp:
```diff
@@ -148,29 +113,27 @@ DenseMatrix transposeDenseMatrix(const DenseMatrix& M)
-                prm_ = setupPropertyTree<TypeTag>(parameters_);
-            }
+            prm_ = setupPropertyTree<TypeTag>(parameters_);
             const auto& gridForConn = simulator_.vanguard().grid();
```
This object (`gridForConn`) is unused now.
